File: skills/academic-figure/templates/rf-0100-regression-diagnostics/r/plot.R
#!/usr/bin/env Rscript

MAX_GROUPS <- 12L
MAX_SAMPLES <- 5000L
POINT_SHAPES <- c(21L, 22L, 24L, 23L, 25L, 7L, 8L, 0L, 2L, 5L, 6L, 1L)
abort <- function(message) stop(message, call. = FALSE)

script_args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", script_args, value = TRUE)
if (length(file_arg) != 1L) abort("Cannot determine plot.R location")
script_dir <- dirname(normalizePath(sub("^--file=", "", file_arg), mustWork = TRUE))
demo <- file.path(dirname(script_dir), "data", "simulated_fixed_seed_supplied_diagnostics.csv")

parse_cli <- function(args) {
  options <- list(input = demo, group_column = NULL, output_prefix = NULL,
                  title = "Supplied regression diagnostics", dpi = "320")
  aliases <- c(input = "input", `group-column` = "group_column", `output-prefix` = "output_prefix",
               title = "title", dpi = "dpi")
  index <- 1L
  while (index <= length(args)) {
    token <- args[[index]]
    if (!startsWith(token, "--")) abort(paste0("Unexpected argument: ", token))
    body <- substring(token, 3L); equals <- regexpr("=", body, fixed = TRUE)[1L]
    if (equals > 0L) {
      key <- substring(body, 1L, equals - 1L); value <- substring(body, equals + 1L)
    } else {
      key <- body; index <- index + 1L
      if (index > length(args) || startsWith(args[[index]], "--")) abort(paste0("Missing value for --", key))
      value <- args[[index]]
    }
    if (!key %in% names(aliases)) abort(paste0("Unknown option --", key))
    if (!nzchar(value)) abort(paste0("Empty value for --", key))
    options[[aliases[[key]]]] <- value; index <- index + 1L
  }
  if (is.null(options$output_prefix)) abort("--output-prefix is required")
  if (grepl("\\.(png|svg|pdf)$", options$output_prefix, ignore.case = TRUE)) abort("--output-prefix must not include an extension")
  if (!grepl("^[0-9]+$", options$dpi)) abort("--dpi must be an integer from 300 to 1200")
  options$dpi <- as.integer(options$dpi)
  if (is.na(options$dpi) || options$dpi < 300L || options$dpi > 1200L) abort("--dpi must be an integer from 300 to 1200")
  options
}

read_input <- function(path) {
  if (!file.exists(path) || dir.exists(path)) abort(paste0("Input file not found: ", path))
  table <- tryCatch(read.csv(path, stringsAsFactors = FALSE, check.names = FALSE, colClasses = "character",
                             na.strings = character(), strip.white = FALSE),
                    error = function(error) abort(paste0("Cannot read input: ", conditionMessage(error))))
  if (nrow(table) < 5L) abort("At least 5 samples are required")
  if (nrow(table) > MAX_SAMPLES) abort(paste0("At most ", MAX_SAMPLES, " samples can be rendered"))
  if (any(!nzchar(trimws(names(table)))) || anyDuplicated(names(table))) abort("Input has empty/duplicate column names")
  missing <- setdiff(c("sample_id", "x", "y", "fitted", "residual"), names(table))
  if (length(missing)) abort(paste0("Input is missing columns: ", paste(missing, collapse = ", ")))
  for (column in names(table)) table[[column]] <- trimws(table[[column]])
  table
}

finite_column <- function(values, name) {
  parsed <- suppressWarnings(as.numeric(values))
  bad <- !nzchar(values) | is.na(parsed) | !is.finite(parsed)
  if (any(bad)) abort(paste0(name, " must contain finite values; invalid CSV lines: ", paste(which(bad) + 1L, collapse = ", ")))
  parsed
}

validate_data <- function(table, group_column) {
  if (any(!nzchar(table$sample_id)) || anyDuplicated(table$sample_id)) abort("sample_id must be non-empty and unique")
  x <- finite_column(table$x, "x"); y <- finite_column(table$y, "y")
  fitted <- finite_column(table$fitted, "fitted"); residual <- finite_column(table$residual, "residual")
  for (item in list(x = x, y = y, fitted = fitted)) if (diff(range(item)) == 0) abort("x, y, and fitted must each vary")
  groups <- NULL
  if (!is.null(group_column)) {
    if (!group_column %in% names(table)) abort(paste0("Input is missing group column '", group_column, "'"))
    groups <- table[[group_column]]
    if (any(!nzchar(groups))) abort("The selected group column cannot contain empty values")
    levels <- unique(groups)
    if (length(levels) > MAX_GROUPS) abort(paste0("The group column exceeds ", MAX_GROUPS, " levels"))
    counts <- table(groups)
    if (any(counts < 2L)) abort("Every group needs at least 2 samples")
  }
  has_status <- "data_status" %in% names(table); has_seed <- "simulation_seed" %in% names(table)
  if (xor(has_status, has_seed)) abort("data_status and simulation_seed must be supplied together")
  if (has_status) {
    provenance <- unique(paste(table$data_status, table$simulation_seed, sep = "\r"))
    if (any(!nzchar(table$data_status)) || any(!nzchar(table$simulation_seed)) || length(provenance) != 1L) abort("Simulation provenance must be non-empty and consistent")
  }
  list(x = x, y = y, fitted = fitted, residual = residual, groups = groups)
}

draw_marginal <- function(values, groups, colors, title, x_label) {
  histogram <- hist(values, breaks = "FD", plot = FALSE)
  curves <- list()
  maximum <- max(histogram$density)
  if (!is.null(groups)) {
    for (level in unique(groups)) {
      selected <- values[groups == level]
      if (length(selected) >= 2L && diff(range(selected)) > 0) {
        curves[[level]] <- density(selected)
        maximum <- max(maximum, curves[[level]]$y)
      }
    }
  }
  plot(histogram, col = "#E5E7EB", border = "white", main = title, xlab = x_label,
       ylab = "Density", freq = FALSE, ylim = c(0, maximum * 1.06))
  for (level in names(curves)) lines(curves[[level]]$x, curves[[level]]$y, col = colors[[level]], lwd = 1.3)
}

draw_plot <- function(data, title) {
  x <- data$x; y <- data$y; fitted <- data$fitted; residual <- data$residual; groups <- data$groups
  levels <- if (is.null(groups)) "All samples" else unique(groups)
  colors <- setNames(grDevices::hcl.colors(length(levels), "Dark 3"), levels)
  shapes <- setNames(POINT_SHAPES[seq_along(levels)], levels)
  group_values <- if (is.null(groups)) rep("All samples", length(x)) else groups
  point_colors <- unname(colors[group_values]); point_shapes <- unname(shapes[group_values])
  layout(matrix(c(1, 2, 4, 1, 3, 4), nrow = 2L, byrow = TRUE), widths = c(4.2, 2.6, 4.2), heights = c(1, 1))
  par(oma = c(4.2, 0.5, 6.0, 0.5))

  par(mar = c(4.2, 4.2, 2.8, 0.8))
  plot(x, y, pch = point_shapes, bg = grDevices::adjustcolor(point_colors, alpha.f = 0.80),
       col = ifelse(point_shapes %in% 21:25, "white", point_colors), xlab = "Supplied x", ylab = "Supplied y / fitted",
       main = "A  Relationship", cex = 0.85)
  grid(col = "#E5E7EB")
  points(x, fitted, pch = 23, bg = NA, col = point_colors, cex = 0.75, lwd = 0.9)
  legend("topleft", legend = c("Supplied y", "Supplied fitted"), pch = c(21, 23), pt.bg = c("#6B7280", NA),
         col = c("white", "#374151"), bty = "n", cex = 0.72)
  if (!is.null(groups)) legend("bottomright", legend = levels, pch = shapes, pt.bg = colors, col = colors,
                               bty = "n", cex = 0.68, ncol = if (length(levels) > 6L) 2L else 1L)

  par(mar = c(3.8, 3.5, 2.8, 0.5))
  draw_marginal(x, groups, colors, "B  Marginal x", "Supplied x")
  par(mar = c(3.8, 3.5, 2.8, 0.5))
  draw_marginal(y, groups, colors, "Marginal y", "Supplied y")

  par(mar = c(4.2, 4.2, 2.8, 0.8))
  plot(fitted, residual, pch = point_shapes, bg = grDevices::adjustcolor(point_colors, alpha.f = 0.80),
       col = ifelse(point_shapes %in% 21:25, "white", point_colors), xlab = "Supplied fitted", ylab = "Supplied residual",
       main = "C  Residual diagnostic", cex = 0.85)
  grid(col = "#E5E7EB"); abline(h = 0, lty = 2, lwd = 1.0)

  mtext(title, side = 3, outer = TRUE, line = 4.2, font = 2, cex = 1.25)
  mtext("All fitted and residual values are supplied; this script fits no model and computes no P values.",
        side = 3, outer = TRUE, line = 2.4, cex = 0.78, col = "#374151")
  mtext("Descriptive visualization only; supplied residuals are not recomputed or statistically verified.",
        side = 1, outer = TRUE, line = 2.2, cex = 0.70, col = "#4B5563")
}

main <- function() {
  options <- parse_cli(commandArgs(trailingOnly = TRUE))
  table <- read_input(options$input); prepared <- validate_data(table, options$group_column)
  levels <- if (is.null(prepared$groups)) character() else unique(prepared$groups)
  width <- max(13.2, 12.8 + 0.04 * max(c(0, nchar(levels, type = "width"))))
  height <- 7.5 + 0.12 * max(0, length(levels) - 4L)
  parent <- dirname(options$output_prefix)
  if (!dir.exists(parent) && !dir.create(parent, recursive = TRUE, showWarnings = FALSE)) abort(paste0("Cannot create output directory: ", parent))
  png_path <- paste0(options$output_prefix, ".png"); pdf_path <- paste0(options$output_prefix, ".pdf")
  png(png_path, width = width, height = height, units = "in", res = options$dpi, bg = "white")
  tryCatch(draw_plot(prepared, options$title), finally = dev.off())
  cairo_pdf(pdf_path, width = width, height = height, bg = "white", onefile = TRUE)
  tryCatch(draw_plot(prepared, options$title), finally = dev.off())
  cat(sprintf("Validated %d unique samples; missing diagnostic values: 0; group levels: %d.\n", nrow(table), length(levels)))
  cat("Supplied fields plotted: x, y, fitted, residual. Models fitted: 0; P values computed: 0.\n")
  cat("Residuals recomputed or statistically verified: no.\n")
  cat("Wrote ", png_path, "\n", sep = ""); cat("Wrote ", pdf_path, "\n", sep = "")
}

tryCatch(main(), error = function(error) { message("ERROR: ", conditionMessage(error)); quit(status = 2L) })

File: skills/academic-figure/templates/rf-0100-regression-diagnostics/python/plot.py
#!/usr/bin/env python3
"""Plot supplied regression values without fitting a model."""

from __future__ import annotations

import argparse
import csv
import math
import sys
from pathlib import Path
from typing import Dict, List, Optional, Sequence, Tuple

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.lines import Line2D


ROOT = Path(__file__).resolve().parents[1]
DEMO = ROOT / "data" / "simulated_fixed_seed_supplied_diagnostics.csv"
MAX_GROUPS = 12
MAX_SAMPLES = 5000
MARKERS = ["o", "s", "^", "D", "v", "P", "X", "<", ">", "h", "p", "*"]


class ValidationError(Exception):
    pass


def read_csv(path: Path) -> Tuple[List[Dict[str, str]], List[str]]:
    if not path.is_file():
        raise ValidationError("Input file not found: {}".format(path))
    try:
        with path.open("r", encoding="utf-8-sig", newline="") as handle:
            reader = csv.DictReader(handle)
            headers = reader.fieldnames
            if headers is None:
                raise ValidationError("Input has no CSV header")
            if any(header is None or not header.strip() for header in headers):
                raise ValidationError("Input contains an empty column name")
            if len(set(headers)) != len(headers):
                raise ValidationError("Input contains duplicate column names")
            required = ["sample_id", "x", "y", "fitted", "residual"]
            missing = [column for column in required if column not in headers]
            if missing:
                raise ValidationError("Input is missing columns: {}".format(", ".join(missing)))
            rows: List[Dict[str, str]] = []
            for line, row in enumerate(reader, start=2):
                if None in row:
                    raise ValidationError("CSV line {} has extra values".format(line))
                clean = {key: (value or "").strip() for key, value in row.items()}
                clean["__line__"] = str(line)
                rows.append(clean)
    except UnicodeDecodeError as exc:
        raise ValidationError("Input must be UTF-8 CSV: {}".format(exc)) from exc
    if len(rows) < 5:
        raise ValidationError("At least 5 samples are required")
    if len(rows) > MAX_SAMPLES:
        raise ValidationError("At most {} samples can be rendered; received {}".format(MAX_SAMPLES, len(rows)))
    return rows, list(headers)


def finite(value: str, context: str) -> float:
    if value == "":
        raise ValidationError("{} cannot be empty".format(context))
    try:
        result = float(value)
    except ValueError as exc:
        raise ValidationError("{} must be numeric; got {!r}".format(context, value)) from exc
    if not math.isfinite(result):
        raise ValidationError("{} must be finite; got {!r}".format(context, value))
    return result


def validate(
    rows: Sequence[Dict[str, str]], headers: Sequence[str], group_column: Optional[str]
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, Optional[List[str]]]:
    if group_column and group_column not in headers:
        raise ValidationError("Input is missing group column {!r}".format(group_column))
    ids: set[str] = set()
    groups: Optional[List[str]] = [] if group_column else None
    values = {name: [] for name in ("x", "y", "fitted", "residual")}
    problems: List[str] = []
    for row in rows:
        line = row["__line__"]
        sample_id = row["sample_id"]
        if not sample_id or sample_id in ids:
            problems.append("line {} has an empty or duplicate sample_id {!r}".format(line, sample_id))
        ids.add(sample_id)
        if groups is not None and group_column is not None:
            group = row[group_column]
            if not group:
                problems.append("line {} has an empty group".format(line))
            groups.append(group)
        for name in values:
            try:
                values[name].append(finite(row[name], "line {} {}".format(line, name)))
            except ValidationError as exc:
                problems.append(str(exc))
                values[name].append(float("nan"))
    if groups is not None:
        levels = list(dict.fromkeys(groups))
        if len(levels) > MAX_GROUPS:
            problems.append("group column has {} levels; maximum is {}".format(len(levels), MAX_GROUPS))
        for level in levels:
            if level and groups.count(level) < 2:
                problems.append("group {!r} has fewer than 2 samples".format(level))
    if problems:
        shown = problems[:25]
        suffix = "\n  - ... additional errors omitted" if len(problems) > 25 else ""
        raise ValidationError("Input validation failed:\n  - " + "\n  - ".join(shown) + suffix)
    arrays = {name: np.asarray(values[name], dtype=float) for name in values}
    for name in ("x", "y", "fitted"):
        if np.ptp(arrays[name]) == 0:
            raise ValidationError("{} must vary across samples".format(name))
    has_status = "data_status" in headers
    has_seed = "simulation_seed" in headers
    if has_status != has_seed:
        raise ValidationError("data_status and simulation_seed must be supplied together")
    if has_status:
        provenance = {(row["data_status"], row["simulation_seed"]) for row in rows}
        if any(not status or not seed for status, seed in provenance) or len(provenance) != 1:
            raise ValidationError("Simulation provenance must be non-empty and consistent")
    return arrays["x"], arrays["y"], arrays["fitted"], arrays["residual"], groups


def render(
    x: np.ndarray,
    y: np.ndarray,
    fitted: np.ndarray,
    residual: np.ndarray,
    groups: Optional[Sequence[str]],
    output_prefix: Path,
    title: str,
    dpi: int,
) -> Tuple[Path, Path]:
    levels = list(dict.fromkeys(groups)) if groups is not None else ["All samples"]
    group_array = np.asarray(groups if groups is not None else ["All samples"] * len(x), dtype=object)
    palette = plt.get_cmap("tab10" if len(levels) <= 10 else "tab20")
    colors = {level: palette(index % palette.N) for index, level in enumerate(levels)}
    max_group_chars = max(len(level) for level in levels)
    width = max(13.2, 12.8 + 0.04 * max_group_chars)
    height = 7.5 + 0.12 * max(0, len(levels) - 4)
    fig = plt.figure(figsize=(width, height))
    outer = fig.add_gridspec(1, 3, width_ratios=[4.3, 2.5, 4.3], left=0.07, right=0.98, bottom=0.13, top=0.83, wspace=0.30)
    ax_relation = fig.add_subplot(outer[0, 0])
    marginal = outer[0, 1].subgridspec(2, 1, hspace=0.38)
    ax_x = fig.add_subplot(marginal[0, 0])
    ax_y = fig.add_subplot(marginal[1, 0])
    ax_residual = fig.add_subplot(outer[0, 2])
    point_size = max(8.0, min(38.0, 1600.0 / len(x)))

    for index, level in enumerate(levels):
        selected = group_array == level
        ax_relation.scatter(x[selected], y[selected], s=point_size, marker=MARKERS[index],
                            facecolor=colors[level], edgecolor="white", linewidth=0.55, alpha=0.80)
        ax_relation.scatter(x[selected], fitted[selected], s=point_size * 0.85, marker="D",
                            facecolor="none", edgecolor=colors[level], linewidth=0.9, alpha=0.90)
        ax_residual.scatter(fitted[selected], residual[selected], s=point_size, marker=MARKERS[index],
                            facecolor=colors[level], edgecolor="white", linewidth=0.55, alpha=0.80)
    ax_relation.set_xlabel("Supplied x")
    ax_relation.set_ylabel("Supplied y / fitted")
    ax_relation.set_title("A  Relationship", loc="left", fontweight="semibold")
    ax_relation.grid(color="#E5E7EB", linewidth=0.65)
    semantic_handles = [
        Line2D([0], [0], marker="o", linestyle="None", markerfacecolor="#6B7280", markeredgecolor="white", label="Supplied y"),
        Line2D([0], [0], marker="D", linestyle="None", markerfacecolor="none", markeredgecolor="#374151", label="Supplied fitted"),
    ]
    ax_relation.legend(handles=semantic_handles, loc="best", frameon=False, fontsize=8)

    bins = max(8, min(24, int(math.ceil(math.sqrt(len(x))))))
    for level in levels:
        selected = group_array == level
        ax_x.hist(x[selected], bins=bins, density=True, histtype="step", linewidth=1.3, color=colors[level])
        ax_y.hist(y[selected], bins=bins, density=True, histtype="step", linewidth=1.3, color=colors[level])
    ax_x.set_title("B  Marginal distributions", loc="left", fontweight="semibold")
    ax_x.set_xlabel("Supplied x")
    ax_y.set_xlabel("Supplied y")
    for axis in (ax_x, ax_y):
        axis.set_ylabel("Density")
        axis.spines["top"].set_visible(False)
        axis.spines["right"].set_visible(False)

    ax_residual.scatter([], [], alpha=0)
    ax_residual.axhline(0, color="#111827", linestyle="--", linewidth=1.0)
    ax_residual.set_xlabel("Supplied fitted")
    ax_residual.set_ylabel("Supplied residual")
    ax_residual.set_title("C  Residual diagnostic", loc="left", fontweight="semibold")
    ax_residual.grid(color="#E5E7EB", linewidth=0.65)
    for axis in (ax_relation, ax_residual):
        axis.spines["top"].set_visible(False)
        axis.spines["right"].set_visible(False)

    if groups is not None:
        handles = [
            Line2D([0], [0], marker=MARKERS[index], linestyle="None", markerfacecolor=colors[level],
                   markeredgecolor="white", markersize=6, label=level)
            for index, level in enumerate(levels)
        ]
        fig.legend(handles=handles, loc="upper center", bbox_to_anchor=(0.5, 0.895),
                   ncol=min(6, len(levels)), frameon=False, fontsize=8.5)
    fig.suptitle(title, y=0.975, fontsize=14, fontweight="semibold")
    fig.text(0.5, 0.94, "All fitted and residual values are supplied; this script fits no model and computes no P values.",
             ha="center", va="top", fontsize=9, color="#374151")
    fig.text(0.5, 0.025, "Descriptive visualization only; supplied residuals are not recomputed or statistically verified.",
             ha="center", va="bottom", fontsize=8, color="#4B5563")
    output_prefix.parent.mkdir(parents=True, exist_ok=True)
    png = Path(str(output_prefix) + ".png")
    svg = Path(str(output_prefix) + ".svg")
    fig.savefig(png, dpi=dpi, facecolor="white", metadata={"Software": "rf-0100 plot.py"})
    fig.savefig(svg, facecolor="white", metadata={"Creator": "rf-0100 plot.py"})
    plt.close(fig)
    return png, svg


def parser() -> argparse.ArgumentParser:
    result = argparse.ArgumentParser(description="Plot supplied regression and residual diagnostics without fitting.")
    result.add_argument("--input", type=Path, default=DEMO)
    result.add_argument("--group-column")
    result.add_argument("--output-prefix", type=Path, required=True)
    result.add_argument("--title", default="Supplied regression diagnostics")
    result.add_argument("--dpi", type=int, default=320)
    return result


def main(argv: Optional[Sequence[str]] = None) -> int:
    args = parser().parse_args(argv)
    if args.dpi < 300 or args.dpi > 1200:
        raise ValidationError("--dpi must be an integer from 300 to 1200")
    if args.output_prefix.suffix.lower() in {".png", ".svg", ".pdf"}:
        raise ValidationError("--output-prefix must not include an extension")
    rows, headers = read_csv(args.input)
    x, y, fitted, residual, groups = validate(rows, headers, args.group_column)
    png, svg = render(x, y, fitted, residual, groups, args.output_prefix, args.title, args.dpi)
    print("Validated {} unique samples; missing diagnostic values: 0; group levels: {}.".format(
        len(x), len(set(groups)) if groups is not None else 0
    ))
    print("Supplied fields plotted: x, y, fitted, residual. Models fitted: 0; P values computed: 0.")
    print("Residuals recomputed or statistically verified: no.")
    print("Wrote {}".format(png))
    print("Wrote {}".format(svg))
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except ValidationError as exc:
        print("ERROR: {}".format(exc), file=sys.stderr)
        sys.exit(2)
